File: js/1. twoNumSum.js
let curentTarget

const array = [3, 5, -4, 8, 11, -1, 1, 6]
const target = 10

// Time Complexity O(N)
// Space Complexity O(N)

const twoNumberSum = (array, target) => {
	const a = {}
	for (let i = 0; i < array.length; i++) {
		const currentElement = array[i]
		const desiredOutput = target - currentElement

		if (a[desiredOutput] !== undefined) {
			return [a[desiredOutput], i]
		}
		a[currentElement] = i
	}
	return []
}

// Time Complexity O(N * log(N))
// Space Complexity O(1)

const twoNumberSum2 = (nums, target) => {
	nums.sort((a, b) => a - b)

	let left = 0
	let right = nums.length - 1

	while (left < right) {
		possibleTarget = nums[left] + nums[right]
		if (target === possibleTarget) return [nums[left], nums[right]]
		if (target < possibleTarget) right -= 1
		if (target > possibleTarget) left += 1
	}

	return []
}

twoNumberSum(array, target)
twoNumberSum2(array, target)

File: python/1. twoNumSum.py
array = [3, 4, 5, 6, 7, 1]
target = 11

# Time Complexity O(N)
# Space Complexity O(N)


def twoNumberSum(nums, target):
    dict = {}
    for num in enumerate(nums):
        sum = target - num
        if sum in dict:
            return [sum, nums[num]]
        else:
            dict[num] = True
    return []


# Time Complexity O(N * log(N))
# Space Complexity O(1)

def twoNumberSum2(nums, target):
    nums.sort()
    left = 0
    right = len(nums) - 1

    while left < right:
        sum = nums[left] + nums[right]
        if sum == target:
            return [nums[left], nums[right]]
        elif sum < target:
            left += 1
        elif sum > target:
            right -= 1

    return []


twoNumberSum(array, target)
twoNumberSum2(array, target)
